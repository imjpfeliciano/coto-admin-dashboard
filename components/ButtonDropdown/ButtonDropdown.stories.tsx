import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import ButtonDropdown from "./ButtonDropdown";

export default {
  title: "Components/ButtonDropdown",
  component: ButtonDropdown,
} as ComponentMeta<typeof ButtonDropdown>;

const Template: ComponentStory<typeof ButtonDropdown> = (args) => (
  <ButtonDropdown {...args} />
);

export const Default = Template.bind({});
Default.args = {
  label: "Click me",
  iconName: "add",
  options: [
    {
      label: "Option 1",
      value: "option1",
    },
    {
      label: "Option 2",
      value: "option2",
    },
    {
      label: "Option 3",
      value: "option3",
    },
  ],
};
