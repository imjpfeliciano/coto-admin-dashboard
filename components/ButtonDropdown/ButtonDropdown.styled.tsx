import styled from "styled-components";
import { BaseButton } from "../Button/Button.styled";

export const DropdownContainer = styled.div`
  position: relative;
`;

export const ButtonControl = styled.div`
  ${BaseButton}

  text-align: center;

  display: flex;
  align-items: center;
  justify-content: space-between;
`;

export const ControlLabelContainer = styled.div`
  display: flex;
  align-items: center;
`;

export const ButtonLabel = styled.span`
  margin-left: 0.5rem;
`;

export const DropdownMenu = styled.div`
  display: block;
  position: absolute;
  top: 100%;
  left: 0;
  right: 0;
  background-color: white;
  z-index: 1;
`;

export const DropdownItem = styled.div`
  padding: 0.5rem 1rem;
  cursor: pointer;
  text-align: left;
  border: 1px solid #ccc;
  border-top: none;
  color: black;

  &:not(:last-of-type) {
    border-bottom: none;
  }

  &:first-of-type {
    border-top-left-radius: 0.5rem;
    border-top-right-radius: 0.5rem;
  }

  &:last-of-type {
    border-bottom-left-radius: 0.5rem;
    border-bottom-right-radius: 0.5rem;
  }

  &:hover {
    background-color: #ddd;
    color: white;
  }
`;
