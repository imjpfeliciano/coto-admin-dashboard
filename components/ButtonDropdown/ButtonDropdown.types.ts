export interface DropdownItemProps {
  id?: string;
  label: string;
  value: string;
}

export interface ButtonDropdownProps {
  label: string;
  iconName: string;
  options: DropdownItemProps[];
  onItemSelect: (item: string) => void;
}
