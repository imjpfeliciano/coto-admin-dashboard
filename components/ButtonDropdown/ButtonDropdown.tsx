import React, { useEffect, useState, useRef } from "react";
import MaterialIcon from "../MaterialIcon";
import {
  DropdownContainer,
  ButtonControl,
  DropdownMenu,
  DropdownItem,
  ControlLabelContainer,
  ButtonLabel,
} from "./ButtonDropdown.styled";
import { ButtonDropdownProps } from "./ButtonDropdown.types";

const ButtonDropdown: React.FC<ButtonDropdownProps> = ({
  label,
  iconName,
  options,
  onItemSelect,
}) => {
  const [isOpen, setIsOpen] = useState(false);
  const ref = useRef();

  useEffect(() => {
    const handleClickOutside = (event) => {
      if (ref.current && !ref.current.contains(event.target)) {
        setIsOpen(false);
      }
    };
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, []);

  return (
    <DropdownContainer ref={ref}>
      <ButtonControl
        onClick={() => setIsOpen(!isOpen)}
        data-testid="button-dropdown"
      >
        <ControlLabelContainer>
          <MaterialIcon iconName={iconName} data-testid="material-icon" />
          <ButtonLabel>{label}</ButtonLabel>
        </ControlLabelContainer>
        <MaterialIcon iconName="expand_more" />
      </ButtonControl>
      {isOpen && (
        <DropdownMenu data-testid="dropdown-menu">
          {options.map((option) => (
            <DropdownItem
              key={option.value}
              onClick={() => {
                onItemSelect(option.value);
                setIsOpen(false);
              }}
            >
              {option.label}
            </DropdownItem>
          ))}
        </DropdownMenu>
      )}
    </DropdownContainer>
  );
};

export default ButtonDropdown;
