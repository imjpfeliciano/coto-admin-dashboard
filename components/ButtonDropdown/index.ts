import ButtonDropdown from './ButtonDropdown'

export default ButtonDropdown
