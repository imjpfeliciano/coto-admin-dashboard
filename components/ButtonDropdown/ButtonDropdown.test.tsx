import ButtonDropdown from "./ButtonDropdown";
import { render, screen, cleanup, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

const MockButtonArgs = {
  label: "Test",
  iconName: "science",
  options: [
    { label: "Option 1", value: "option1" },
    { label: "Option 2", value: "option2" },
  ],
  onItemSelect: jest.fn(),
};

describe("ButtonDropdown", () => {
  afterEach(() => {
    cleanup();
  });

  it("should render", () => {
    const { container } = render(<ButtonDropdown {...MockButtonArgs} />);
    expect(container).toMatchSnapshot();
  });

  it("should render with label", async () => {
    render(<ButtonDropdown {...MockButtonArgs} />);
    expect(screen.findByText("Test")).toBeTruthy();
  });

  it("should render with icon", async () => {
    render(<ButtonDropdown {...MockButtonArgs} />);
    // FIXME: Check why .toBeInTheDocument() is not working
    expect(screen.getByTestId("material-icon")).toBeTruthy();
  });

  it("should open the options menu", async () => {
    render(<ButtonDropdown {...MockButtonArgs} />);
    const button = screen.getByTestId("button-dropdown");
    await userEvent.click(button);
    expect(screen.getByTestId("dropdown-menu")).toBeTruthy();
  });

  it("should close the options menu", async () => {
    render(<ButtonDropdown {...MockButtonArgs} />);
    const button = screen.getByTestId("button-dropdown");
    await userEvent.click(button);
    await userEvent.click(button);
    expect(screen.queryByTestId("dropdown-menu")).not.toBeTruthy();
  });

  it("should call onItemSelect when an option is clicked", async () => {
    render(<ButtonDropdown {...MockButtonArgs} />);
    const button = screen.getByTestId("button-dropdown");
    await userEvent.click(button);
    const option = screen.getByText("Option 1");

    await userEvent.click(option);
    expect(MockButtonArgs.onItemSelect).toHaveBeenCalledWith("option1");
  });
});
