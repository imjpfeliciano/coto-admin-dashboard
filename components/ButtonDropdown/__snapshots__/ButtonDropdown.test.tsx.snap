// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonDropdown should render 1`] = `
<div>
  <div
    class="ButtonDropdownstyled__DropdownContainer-sc-1ket5fv-0 bozSIJ"
  >
    <div
      class="ButtonDropdownstyled__ButtonControl-sc-1ket5fv-1 ehNdEF"
      data-testid="button-dropdown"
    >
      <div
        class="ButtonDropdownstyled__ControlLabelContainer-sc-1ket5fv-2 gmvqdd"
      >
        <span
          class="MaterialIcon__IconContainer-sc-1s3bnun-0 fVqaYP material-icons"
          data-testid="material-icon"
        >
          science
        </span>
        <span
          class="ButtonDropdownstyled__ButtonLabel-sc-1ket5fv-3 dgqcGq"
        >
          Test
        </span>
      </div>
      <span
        class="MaterialIcon__IconContainer-sc-1s3bnun-0 fVqaYP material-icons"
      >
        expand_more
      </span>
    </div>
  </div>
</div>
`;
